### PR TITLE
1.7 release announcement

### PR DESCRIPTION
**Pull request by steveklabnik**

### CODE REVIEW

File: _posts/2016-03-02-Rust-1.7.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+### What's in 1.7 stable
+
+This release contains a number of small refinements, one major feature, and
+a change to [Crates.io](https://crates.io).
+
+#### New Language features
+
+... there were none! While we have several things cooking for future releases,
+the timeframe in which 1.7 included the holidays, which means less time for
```
Comment (nikomatsakis) [inline _posts/2016-03-02-Rust-1.7.md]: not sure if we need an excuse here. Maybe just "for future releases, none of them landed for the 1.7 cutoff."

File: _posts/2016-03-02-Rust-1.7.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+commenting on GitHub and more time for spending with loved ones.
```
Comment (nikomatsakis) [inline _posts/2016-03-02-Rust-1.7.md]: thought I do like talking about your sweet cuddly loved ones :)

File: _posts/2016-03-02-Rust-1.7.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+As always, you can [install Rust 1.7][install] from the appropriate page on our
+website, and check out the [detailed release notes for 1.7][notes] on GitHub.
+About 1300 patches were landed in this release.
+
+[install]: http://www.rust-lang.org/install.html
+[notes]: https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-170-2016-03-03
+
+### What's in 1.7 stable
+
+This release contains a number of small refinements, one major feature, and
```
Comment (nikomatsakis) [inline _posts/2016-03-02-Rust-1.7.md]: What is the major feature? :) 

Maybe we should just say something like: "This release is primarily about library stabilization. We stabilized approximately 40 new library functions and methods, including: blah blah blah. 

There are no new language features in this release, though we do have several things cooking for future releases."

Comment (steveklabnik) [inline _posts/2016-03-02-Rust-1.7.md]: This is just a typo :sob:

File: _posts/2016-03-02-Rust-1.7.md (hunk 2 above)
Comment (steveklabnik) [inline _posts/2016-03-02-Rust-1.7.md]: yeah, it just seemed like an interesting way to acknowledge the train model. We can always cut it.

File: _posts/2016-03-02-Rust-1.7.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+[notes]: https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-170-2016-03-03
+
+### What's in 1.7 stable
+
+This release is primarily about library features. While we have several
+language features cooking for future releases, the timeframe in which 1.7
+included the holidays, which means less time for commenting on GitHub and more
+time for spending with loved ones.
+
+#### New Language features
```
Comment (aturon) [inline _posts/2016-03-02-Rust-1.7.md]: Let's just drop this section.

File: _posts/2016-03-02-Rust-1.7.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+`None`
+
+#### Library stabilizations
+
+About 40 library functions and methods are now stable in 1.7. Notable
+improvements include:
+
+* The ability to customize `HashMap<K, V>`s and similar data structures with
```
Comment (aturon) [inline _posts/2016-03-02-Rust-1.7.md]: Maybe we can elaborate on this point a bit -- @alexcrichton, maybe you could write up a paragraph or two about how custom hashing works?

Comment (aturon) [inline _posts/2016-03-02-Rust-1.7.md]: In particular, this is a good opportunity to talk about how our hashing design differs from the simple story in e.g. Java, and what the implications are.

**Comment by alexcrichton** [comment]
This was a pretty quiet release for Cargo as well, but there were a few PRs we may wish to highlight:

* https://github.com/rust-lang/cargo/pull/2279 - this was a relatively large change, and it should allow build scripts to precisely inform Cargo about dependencies to ensure that they're only rerun when those files change. This should help development quite a bit in repositories with build scripts.
* https://github.com/rust-lang/cargo/pull/2224 - a modification to the `cargo rustc` subcommand, allows specifying profiles to pull in dev-dependencies during testing and such.

File: _posts/2016-03-02-Rust-1.7.md (hunk 5 above)
Comment (alexcrichton) [inline _posts/2016-03-02-Rust-1.7.md]: Sure yeah, perhaps something like (wordsmithing always welcome):

---

One of the largest APIs stabilized during the 1.7 cycle was support for custom hash algorithms in the standard library's `HashMap` type. Previously all hash maps would use [SipHash](https://en.wikipedia.org/wiki/SipHash) as the hashing algorithm, which provides protection against DOS attacks by default. SipHash, however, is [not very fast](http://cglab.ca/~abeinges/blah/hash-rs/) at hashing small keys. As shown, however, the [FNV hash algorithm](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function) is much faster for these size of inputs. This means that by switching hash algorithms for types like `HashMap<usize, V>` there can be a significant speedup so long as the loss of DOS protection is acceptable. 

To see this in action, you can check out the [fnv crate](https://crates.io/crates/fnv) on crates.io and create a `HashMap` via:

```rust
extern crate fnv;

use std::collections::HashMap;                            
use std::hash::BuildHasherDefault;                        
use fnv::FnvHasher;                                       
                                                          
type MyHasher = BuildHasherDefault<FnvHasher>;            
                                                          
let mut map: HashMap<_, _, MyHasher> = HashMap::default();
map.insert(1, "Hello");                                   
map.insert(2, ", world!");                                
println!("{:?}", map);
```

Note that most of the time you don't even need to specify the hasher as type inference will take care of it, so `HashMap::default()` should be all you need to get up to 2x faster hashes. It's also worth pointing out that [`Hash` trait](http://doc.rust-lang.org/std/hash/trait.Hash.html) is agnostic to the hashing algorithm used, so no changes are needed to the types being inserted into hash maps to reap the benefits!

Comment (aturon) [inline _posts/2016-03-02-Rust-1.7.md]: :+1:, very nice!

**Comment by steveklabnik** [comment]
Updates pushed. Thanks a ton @alexcrichton .

File: _posts/2016-03-02-Rust-1.7.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+---
+
+The Rust team is happy to announce the latest version of Rust, 1.7. Rust is a
+systems programming language focused on safety, speed, and concurrency.
+
+As always, you can [install Rust 1.7][install] from the appropriate page on our
+website, and check out the [detailed release notes for 1.7][notes] on GitHub.
+About 1300 patches were landed in this release.
+
+[install]: http://www.rust-lang.org/install.html
```
Comment (frewsxcv) [inline _posts/2016-03-02-Rust-1.7.md]: nit: `https://`

File: _posts/2016-03-02-Rust-1.7.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+website, and check out the [detailed release notes for 1.7][notes] on GitHub.
+About 1300 patches were landed in this release.
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-170-2016-03-03
+
+### What's in 1.7 stable
+
+This release is primarily about library features. While we have several
+language features cooking for future releases, the timeframe in which 1.7
```
Comment (alexcrichton) [inline _posts/2016-03-02-Rust-1.7.md]: "in which 1.7 was developed"

Comment (steveklabnik) [inline _posts/2016-03-02-Rust-1.7.md]: :heart:

**Pull request merged**